src/components/Secrets/SecretForm.tsx

```diff
@@ -1,32 +1,60 @@
-import React from 'react';
+import React, { useEffect } from 'react';
 import { Form } from '@patternfly/react-core';
 import { SelectVariant } from '@patternfly/react-core/deprecated';
 import { useFormikContext } from 'formik';
 import { DropdownItemObject, SelectInputField } from '../../shared';
 import KeyValueFileInputField from '../../shared/components/formik-fields/key-value-file-input-field/KeyValueFileInputField';
-import { SecretFormValues, SecretTypeDropdownLabel } from '../../types';
+import {
+  SecretFormValues,
+  SecretTypeDropdownLabel,
+  K8sSecretType,
+  ExistingSecret,
+} from '../../types';
 import { RawComponentProps } from '../modal/createModalLauncher';
 import SecretTypeSelector from './SecretTypeSelector';
 import {
+  supportedPartnerTasksSecrets,
   getSupportedPartnerTaskKeyValuePairs,
   isPartnerTask,
-  getSupportedPartnerTaskSecrets,
 } from './utils/secret-utils';
 
 type SecretFormProps = RawComponentProps & {
-  existingSecrets: string[];
+  existingSecrets: ExistingSecret[];
 };
 
 const SecretForm: React.FC<React.PropsWithChildren<SecretFormProps>> = ({ existingSecrets }) => {
   const { values, setFieldValue } = useFormikContext<SecretFormValues>();
+  const [currentType, setType] = React.useState(values.type);
   const defaultKeyValues = [{ key: '', value: '', readOnlyKey: false }];
   const defaultImageKeyValues = [{ key: '.dockerconfigjson', value: '', readOnlyKey: true }];
+  const [options, setOptions] = React.useState([]);
+  const [optionsValues, setOptionsValues] = React.useState(null);
 
-  const initialOptions = getSupportedPartnerTaskSecrets().filter(
-    (secret) => !existingSecrets.includes(secret.value),
-  );
-  const [options, setOptions] = React.useState(initialOptions);
-  const currentTypeRef = React.useRef(values.type);
+  useEffect(() => {
+    const initialOptions = existingSecrets
+      .filter((secret) => secret.type === K8sSecretType[currentType])
+      .concat(
+        currentType === SecretTypeDropdownLabel.opaque &&
+          existingSecrets.find((s) => s.name === 'snyk-secret') === undefined
+          ? [supportedPartnerTasksSecrets.snyk]
+          : [],
+      )
+      .filter((secret) => secret.type !== K8sSecretType[SecretTypeDropdownLabel.image])
+      .map((secret) => ({ value: secret.name, lable: secret.name }));
+    const initialOptionsValues = existingSecrets
+      .filter((secret) => secret.type === K8sSecretType[currentType])
+      .filter((secret) => secret.type !== K8sSecretType[SecretTypeDropdownLabel.image])
+      .reduce(
+        (dictOfSecrets, secret) => {
+          dictOfSecrets[secret.name] = secret;
+          return dictOfSecrets;
+        },
+        { 'snyk-secret': supportedPartnerTasksSecrets.snyk },
+      );
+
+    setOptions(initialOptions);
+    setOptionsValues(initialOptionsValues);
+  }, [currentType, existingSecrets]);
```

Comment:
You don't need useEffect and useState here, Just use useMemo to get the values you need since you are just traversing the values.

Reply:
Thanks @sahil143 for the tip! I did not know that and it worked. New version using `useMemo` and without state.